hazelcast/src/main/java/com/hazelcast/util/ServiceLoader.java

```diff
@@ -338,32 +342,106 @@ public void remove() {
             iterator = serviceDefinitions.iterator();
+            this.expectedType = expectedType;
         }
 
         @Override
         public boolean hasNext() {
-            return iterator.hasNext();
+            if (nextClass != null) {
+                return true;
+            }
+            return advance();
         }
 
-        @Override
-        @SuppressWarnings("unchecked")
-        public Class<T> next() {
-            ServiceDefinition definition = iterator.next();
-            try {
+        private boolean advance() {
+            while (iterator.hasNext()) {
+                ServiceDefinition definition = iterator.next();
                 String className = definition.className;
                 ClassLoader classLoader = definition.classLoader;
-                return (Class<T>) ClassLoaderUtil.loadClass(classLoader, className);
-            } catch (Exception e) {
+
+                try {
+                    Class<?> candidate = loadClass(className, classLoader);
+                    if (expectedType.isAssignableFrom(candidate)) {
+                        nextClass = (Class<T>) candidate;
+                        return true;
+                    } else {
+                        onNonAssignableClass(className, candidate);
+                    }
+                } catch (ClassNotFoundException e) {
+                    onClassNotFoundException(className, classLoader, e);
+                }
+            }
+            return false;
+        }
+
+        private Class<?> loadClass(String className, ClassLoader classLoader) throws ClassNotFoundException {
+            Class<?> candidate;
+            if (USE_CLASSLOADING_FALLBACK) {
+                candidate = ClassLoaderUtil.loadClass(classLoader, className);
+            } else {
+                candidate = classLoader.loadClass(className);
+            }
+            return candidate;
+        }
+
+        private void onClassNotFoundException(String className, ClassLoader classLoader, ClassNotFoundException e) {
+            if (className.startsWith("com.hazelcast")) {
+                LOGGER.fine("Failed to load " + className + " by " + classLoader
+                        + ". This indicates a classloading issue. It can happen in a runtime with "
+                        + "a complicated classloading model. (OSGi, Java EE, etc);");
+            } else {
                 throw new HazelcastException(e);
             }
         }
 
+        private void onNonAssignableClass(String className, Class candidate) {
+            if (expectedType.isInterface()) {
+                if (ClassLoaderUtil.implementsInterfaceWithSameName(candidate, expectedType)) {
+                    // this can happen in application containers - different Hazelcast JARs are loaded
+                    // by different classloaders.
+                    LOGGER.fine("There appears to be a classloading conflict. "
```

Comment:
I am not sure. it's probably a business as usual when deployed an app. container. 
it provides no value unless you are troubleshooting it.